lib/hsa/activity_prof.h

```diff
@@ -0,0 +1,175 @@
+THE SOFTWARE.
+*/
+
+#ifndef ACTIVITY_PROF_H
+#define ACTIVITY_PROF_H
+
+#include <mutex>
+
+#include "kalmar_runtime.h"
+
+#if USE_PROF_API
+#include "prof_protocol.h"
+#include "hsa_rt_utils.hpp"
+
+#define ACTIVITY_PROF_INSTANCES()                                                    \
+    namespace activity_prof {                                                        \
+        CallbacksTable::table_t CallbacksTable::_table{};                            \
+        CallbacksTable::mutex_t CallbacksTable::_mutex;                              \
+        template<> HSAOp::ActivityProf::timer_t* HSAOp::ActivityProf::_timer = NULL; \
+        template<> std::atomic<record_id_t> HSAOp::ActivityProf::_glob_record_id(0); \
+    } // activity_prof
+
+namespace activity_prof {
+typedef uint64_t record_id_t;
+typedef uint32_t op_id_t;
+typedef uint32_t activity_kind_t;
+typedef Kalmar::hcCommandKind command_id_t;
+
+typedef activity_id_callback_t id_callback_fun_t;
+typedef activity_async_callback_t callback_fun_t;
+typedef void* callback_arg_t;
+
+class CallbacksTable {
+    public:
+    struct table_t {
+        id_callback_fun_t id_callback;
+        callback_fun_t fun[hc::HSA_OP_ID_NUM];
+        callback_arg_t arg[hc::HSA_OP_ID_NUM];
+    };
+    typedef std::recursive_mutex mutex_t;
+
+    static void set_id_callback(const id_callback_fun_t& fun) { _table.id_callback = fun; }
+    static id_callback_fun_t get_id_callback() { return _table.id_callback; }
+    
+    static bool set_async_callback(const op_id_t& op_id, const callback_fun_t& fun, const callback_arg_t& arg) {
+        std::lock_guard<mutex_t> lck(_mutex);
+        if (op_id == hc::HSA_OP_ID_ANY) {
+            for (op_id_t i = 0; i < hc::HSA_OP_ID_NUM; ++i) {
+                set_async_callback(i, fun, arg);
+            }
+        } else if (op_id < hc::HSA_OP_ID_NUM) {
+            _table.fun[op_id] = fun;
+            _table.arg[op_id] = arg;
+        } else {
+          return false;
+        }
+        return true;
+    }
+    
+    static void get_async_callback(const op_id_t& op_id, callback_fun_t* fun, callback_arg_t* arg) {
+        std::lock_guard<mutex_t> lck(_mutex);
+        *fun = _table.fun[op_id];
+        *arg = _table.arg[op_id];
+    }
+
+    private:
+    static table_t _table;
+    static mutex_t _mutex;
+};
+
+// Activity profile class
+template <typename OpCoord>
+class ActivityProf {
+public:
+    // Domain ID
+    static const int ACTIVITY_DOMAIN_ID = ACTIVITY_DOMAIN_HCC_OPS;
+    // Timeer type
+    typedef hsa_rt_utils::Timer timer_t;
+
+    ActivityProf(const op_id_t& op_id, const OpCoord& op_coord) :
+        _op_id(op_id),
+        _op_coord(op_coord),
+        _callback_fun(NULL),
+        _callback_arg(NULL),
+        _record_id(0)
+    {}
+
+    // Initialization
+    void initialize() {
+        CallbacksTable::get_async_callback(_op_id, &_callback_fun, &_callback_arg);
+        if (_callback_fun != NULL) {
```

Comment:
The problem that you can enable tracing of specific OPs

Reply:
I still prefer having an `enabled` flag.  It could be one per OP.  It makes the code more explicit about what it's doing.
The callback function is always the same for a given during the entire lifetime of the application?

Reply:
It's for runtime library instance as CUPTI record data has thread id and process id

Reply:
Activity tracing can be enabled and disabled in any moment and it is enabled for runtime library instance as CUPTI record data has thread id and process id.

Reply:
Will add the flag.